File: src/main/java/site/liangbai/realhomehunt/listener/player/block/ListenerBlockSpread.java
/*
 * RealHomeHunt
 * Copyright (C) 2021  Liangbai
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package site.liangbai.realhomehunt.listener.player.block;

import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockSpreadEvent;
import site.liangbai.dynamic.event.EventSubscriber;
import site.liangbai.realhomehunt.api.residence.Residence;
import site.liangbai.realhomehunt.api.residence.attribute.impl.SpreadFireAttribute;
import site.liangbai.realhomehunt.api.residence.manager.ResidenceManager;

@EventSubscriber
public final class ListenerBlockSpread implements Listener {
    @EventHandler
    public void onBlockSpread(BlockSpreadEvent event) {
        Block block = event.getBlock();

        if (block.getType() != Material.FIRE) return;

        if (!ResidenceManager.isOpened(block.getWorld())) return;

        Residence residence = ResidenceManager.getResidenceByLocation(block.getLocation());

        if (residence != null) {
            if (!residence.checkBooleanAttribute(SpreadFireAttribute.class)) event.setCancelled(true);
        }
    }
}

File: RealHomeHunt/src/main/java/site/liangbai/realhomehunt/internal/listener/player/block/ListenerBlockSpread.kt
/*
 * RealHomeHunt
 * Copyright (C) 2022  Liangbai
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package site.liangbai.realhomehunt.internal.listener.player.block

import org.bukkit.Material
import org.bukkit.event.block.BlockSpreadEvent
import site.liangbai.realhomehunt.api.residence.attribute.impl.SpreadFire
import site.liangbai.realhomehunt.api.residence.manager.ResidenceManager
import taboolib.common.platform.event.SubscribeEvent

internal object ListenerBlockSpread {
    @SubscribeEvent
    fun onBlockSpread(event: BlockSpreadEvent) {
        val block = event.block
        if (block.type != Material.FIRE) return
        if (!ResidenceManager.isOpened(block.world)) return
        val residence = ResidenceManager.getResidenceByLocation(block.location)
        if (residence != null) {
            if (!residence.checkBooleanAttribute<SpreadFire>()) event.isCancelled = true
        }
    }
}